File: searches/linear_search.cpp
#include <iostream>
#include <vector>

/**
 * @brief Linear Search implementation in C++
 *
 * @param arr Vector to search
 * @param target Value to search for
 * @return int Index of target or -1 if not found
 */
int linear_search(const std::vector<int> &arr, int target) {
  for (size_t i = 0; i < arr.size(); ++i) {
    if (arr[i] == target) {
      return i;
    }
  }
  return -1;
}

int main() {
  std::vector<int> arr = {10, 50, 30, 70, 80, 20};
  int target = 30;

  int result = linear_search(arr, target);

  if (result != -1)
    std::cout << "Element found at index " << result << std::endl;
  else
    std::cout << "Element not found" << std::endl;

  return 0;
}

File: searches/linear_search.rs
/**
 * Linear Search implementation in Rust
 */

pub fn linear_search<T: PartialEq>(arr: &[T], target: &T) -> Option<usize> {
    for (index, item) in arr.iter().enumerate() {
        if item == target {
            return Some(index);
        }
    }
    None
}

fn main() {
    let arr = vec![10, 50, 30, 70, 80, 20];
    let target = 30;
    
    match linear_search(&arr, &target) {
        Some(index) => println!("Element found at index {}", index),
        None => println!("Element not found"),
    }
}
